eclair-core/src/main/scala/fr/acinq/eclair/wire/FailureMessage.scala

```diff
@@ -29,7 +31,8 @@ import scodec.Attempt
 
 // @formatter:off
 sealed trait FailureMessage { def message: String }
-sealed trait BadOnion extends FailureMessage { def onionHash: ByteVector32 }
+sealed trait FailureCode { def failureCode: Int }
```

Comment:
Since all `failure_message`s have a `failure_code`, it should be:

```scala
sealed trait FailureMessage {
  def code: Int
  def message: String
}
```
But then we have to define `failureCode` for each message, right? And then it gets nasty...

Reply:
Yeah exactly, the reason we have to do this for onion errors is that the spec is a bit messy on that.
The only place where we really need that mapping in the other direction is for BadOnion because it's used in FailureMessage and in the malformed onion message.
The other FailureMessages don't need that "reverse lookup" so I didn't bother doing it for all.
But it wouldn't be a lot of work, if you want I can harmonize this.

Reply:
The problem is that the result would probably not look good, we would have to define companion objects to all classes (or use the `FailureMessage` object to store the bindings).

Reply:
Ah that's true.
Let me investigate a bit though, because I think that `BadOnion` errors should in fact be completely separated from `FailureMessage` which would fix this problem.